### ISSUE TITLE
localforage-lite ?

### ISSUE THREAD

**Issue by pastelsky** [localForage/localForage#546]
The library does a brilliant work of abstracting client side storage using localStorage, WebSQL and IndexDB, but the js file size might be a limitation for many-many projects. 

This library bundles multiple drivers and API design options (Node style vs. Promises) into a single file, and doesn't really provide any option to the developer to chose files he wants to include at runtime. 
eg. Not all devs would want to bundle both localStorage and WebSQL, since they have similar browser support.

It would be awesome if `localforage` by itself was a minimal API wrapper, and drivers / API options could be bundled as separate addons. Modular Js is the next big thing after-all.

**Comment by thgreasi** [comment]
What's your suggestion about the way the the various parts would be
included?

Atm such a modular architecture with requirejs (loading only the requested
drivers) is present but deprecated. A similar pattern can work for global
imports.
Perhaps we could also have a separate  es5 dist build, where each driver
could be available on its own and localforage.js would only contain the API
wrapper.

On the other hand, I think that localforage is exploring a way to provide
only a single solid dist file with all the drivers included. Currently #538
drops the not-yet-in-es-spec System.import() in favor of pure es6 imports,
which produce a single solid output file and I think that it conflicts with
this request. In that approach it's difficult to have the proposed feature,
except if a custom build architecture is also implemented.

TL;DR + final suggestion: We could start using the custom drivers
architecture for all the drivers and demand from the developers to include
them. This could be a separate build from the existing all-in-one dist file
and it also doesn't conflict with what #538 achives!


ps: sorry if writing all my thought tired you

> --
*Thodoris Greasidis*
Computer, Networking &
Communications Engineer

**Comment by shubhamsizzles** [comment]
In effect it should be similar to how React JS allows one to import add-ons through a UMD export, or how lodash exports each of its utlilities. Also, it would be great if it ES6 style imports were supported too. A lot of people have been wishing for  a plug and play system in localforage.

**Comment by tofumatt** [comment]
This sounds like something we would need to build a "localForage packager" for; some kind of site that built you a custom version of localForage with only the drivers you wanted.

Omitting promises/callbacks would be more complex I'd think, I'm not sure worth the effort... but if done sanely I'm open to both these ideas. :+1:

**Comment by thgreasi** [comment]
@shubhamsizzles what do you mean by "**A lot of people have been wishing for a plug and play system in localforage.**"? Don't we already have that?

**Comment by nolanlawson** [comment]
Well, there's a PR open that shrinks the bundle size by .5KB (down to 7.7KB). :wink: https://github.com/mozilla/localForage/pull/538

That said, I'd argue that the current file size (<10KB) is already darn small. I think making it more granular than that is really just splitting hairs. `react-dom` alone is 55KB; we're already well under that.

Furthermore, the current landscape of browser storage makes it unfeasible (in my opinion) to split up the adapters and allow users to pick their own. There are subtleties in the usability of e.g. IndexedDB on WKWebView in Chrome for iOS that mean the situation is enormously complicated, and so expecting users to "choose their own" is an invitation to have LocalForage break in an odd browser, all to save ~2KB or so.

**Comment by tofumatt** [comment]
Shoot, you’re right. I started moving out of my flat and other stuff and became crazy distracted. But I scheduled a week this month to dedicate solely to localForage, so I’ll have that in soon!

**Comment by thgreasi** [comment]
:+1: @nolanlawson 
We should start reviewing (and merge) #538 .
From that point and after, as a starting point , we could also split the `import`s of the default drivers into a separate module.

**Comment by shubhamsizzles** [comment]
@nolanlawson You're only comparing gzipped size differences, that too with all storage options enabled (if i'm right that is). 

It may be a 3-4KB decrement, but then when one has a production app with large catalog of libraries, every KB counts. Javascript engines on mobile devices (where using localforage makes a lot more sense) are relatively slow, and its also about how much extra JS you are parsing vs. how much benefit you get out of it.

Moreover, the current localforage configuration already allows you to pick your storage options, so I don't see any problem in doing that.

**Comment by shubhamsizzles** [comment]
@thgreasi  With "plug and play" I was referring to the "drop in the drivers you need" approach.
A little bit like how [`redux-replicate`](https://www.npmjs.com/package/redux-replicate) allows one to store  application state using multiple drivers like `redux-replicate-localforage` and `redux-replicate-fs` etc.

**Comment by shubhamsizzles** [comment]
In my opinion this library can be broken down to:

1. **localforage** : The minimal API wrapper that binds everything together.
2. **localforage-driver-localstorage**: Add support for localStorage.
3. **localforage-driver-websql**: Add support for websql.
4. **localforage-serializer**: Useful for storing binary data for websql & localStorage (only).
4. **localforage-driver-indexeddb**: Add support for indexeddb.

The following can be removed:
1. NodeJs style callbacks: Now with even node moving towards harmony, callbacks are going to be a thing of the past.
2. Promise dependency:   Many projects already add their own polyfill for Promise, this would just be a duplication. Plus it gives freedom to use their choice of a  standards compliant Promise implementation.

**Comment by tofumatt** [comment]
Keep in mind that localForage's primary aim is to be a drop-in, easy-to-use library. The API is _very_ basic and not powerful. It's easier to use than IndexedDB, but by design it has a lot of code that's beyond just a simple IDB wrapper with Promises, which it sounds like what you're after.

Removing the node-style callbacks was discussed but we're not there yet. When callbacks _are_ a thing of the past we'll remove them, but we aren't there yet (sadly).

We do include a no-promises library, fwiw.

**Comment by shubhamsizzles** [comment]
@tofumatt  Agreed, it can still ship with a easy-to-use dist version that bakes all options into one (ReactJS does ship with one too). 
But any plans of modularizing this thing into multiple npm packages for power users would be great.

**Comment by nolanlawson** [comment]
@shubhamsizzles I tend to agree with @tofumatt that this library should remain drop-in and easy-to-use. One of the things scaring people away from browser storage (IMO) is that it's so easy to write something that doesn't work in browser X. LocalForage solves that problem.

I'd say if you're comfortable defining your own adapter system and managing your own browser support, you may want to look into something very lite like https://github.com/jakearchibald/indexeddb-promised and then using https://github.com/axemclion/IndexedDBShim if you need WebSQL support as well.

@thgreasi Indeed, once the ES6 modules are in, you can expose a `jsnext:main` and then allows folks to "build your own" localforage using simple `import`s. :) Then tools like Rollup and Webpack 2 will tree-shake automatically. Still, I think it's a niche use case.

**Comment by shubhamsizzles** [comment]
@nolanlawson 
The indexedDBshim library is in itself twice the size of localForage with all adapters.

A good number of projects already use Webpack  / Rollup. (ReactJS/AngularJS community for eg.) While the current implementation of webpack doesn't support tree shaking, hypothetically, one can still directly reference imports like say, `import WebSQLDriver from localforage/lib/WebSQL` and pick up only the relevant parts (similar to how one can pick up individual functions from lodash without including the whole library). 

So while localForage with all adapters can continue to be the default export, power users can play around with what they need.

**Comment by nolanlawson** [comment]
I suggest revisiting this once Node has figured out how it wants to handle ES6 modules (`.mjs`? `jsnext:main`?) and then see how easy it would be to `export` a modular LocalForage. Publishing as separate modules is probably pretty bonkers for a library that is only 7.6KB (each adapter seems to be about 2KB, don't see how the promise/callback stuff could easily be extracted out).

I'll also point out that while, yes, Lodash and React do factor out their codebase to an extreme amount, it's because their codebases are very large. By my measure, `lodash.map` is 5664 bytes, `react-dom` is 60794 bytes, whereas the master branch of LocalForage is 7685 bytes.

**Comment by tofumatt** [comment]
>  The indexedDBshim library is in itself twice the size of localForage with all adapters.

I think that speaks to localForage's reasonably small size 😄 

Thinking this over again, I don't think we need to refactor and modularise anymore for max a few kB of possibly unused drivers. localForage is about broad browser support and ease-of-use, but could easily be adapted to be used without all drivers.

I think localForage is slim enough, especially after #538 landed. This level of modularisation seems overkill when there are more pressing issues regarding performance to explore (we still need to land batch writes).

**Comment by nolanlawson** [comment]
Nice post on why multi-package monorepos are really hard to achieve: https://medium.com/@jonathanewerner/thoughts-about-package-modularization-d9631f7a41f1

**Comment by lukepighetti** [comment]
Just wanted to comment that comparing to React for what is an acceptable file size is not always appropriate, since some teams select Preact (3kB) when targeting emerging economies with progressive web apps.

**Comment by thgreasi** [comment]
Just an update for this. Now that we use `defineDriver()` internally for the internal drivers, I'm thinking about a way that consumers could only import the drivers they want and instantiate localforage with them. I would expect this to be part of v2.0 .

**Comment by tofumatt** [comment]
We've been using [lerna](https://lernajs.io) in @WordPress/gutenberg (https://github.com/WordPress/gutenberg/blob/master/lerna.json) as well to manage multiple packages in one repo. Maybe we just start publishing drivers as npm packages and see about letting people build there own by importing their own packages 🤔

**Comment by thgreasi** [comment]
In my opinion lerna might be an overkill for this (subjective opinion).
We have a few common utilities that would require some thinking while splitting the modules and are source files are not so loosly coupled atm anyway. Another concern could be that having an additional step in our process could hurt contributions b/c.
Maybe we could for start try something simple, similarly to how lodash allows us to import the individual submodules we want.

If on the other hand you still think that it would be easy and it's just me who feels a bit weird about lerna (or maybe I'm afraid of the change), then I'm totally fine with it and trust you on this :+1:

**Comment by lukepighetti** [comment]
I have definitely found that the ability to import specific modules at a lower level is very useful. For example, I can import `import { BehaviorSubject } from "rxjs";` or `import { BehaviorSubject } from "rxjs/internal/BehaviorSubject";`

I have to admit I have never tested the difference on package size, but assuming it can be reduced with this type of method I think its fine.

But I am only one opinion and I'm pretty new to web tech so, take it with a grain of salt.

**Comment by tony-scio** [comment]
Just curious if there are still plans here.

I think there's a simple thing that could be a big win. That's to just move the `LocalForage` class (https://github.com/localForage/localForage/blob/master/src/localforage.js#L85) to a standalone file. Then a user could import only the indexeddb driver and the LocalForage class and configure it to drop support for localstorage and websql.

For many projects (including mine) all browsers we care about support indexeddb:
https://caniuse.com/#feat=indexeddb

And webpack can prune away the dead code if we can just import what we use.

**Comment by tofumatt** [comment]
We were discussing it in #984, actually.

I think it's still a good idea, but no immediate plans. I'm working a full-scale rewrite in TypeScript, which should be more modular by nature and allow for this sort of thing. I'd imagine shortly after that rewrite is complete we'll publish packages for the "core" and "driver" modules, allowing exactly what you describe.

Sadly no ETA, but it's something I'd like to do 😄
